### PR TITLE
External control plane topology

### PR DESCRIPTION
**Pull request by csrwng**
Add support for indicating whether a control plane is external through the ControlPlaneTopology field

### CODE REVIEW

**Comment by dhellmann** [comment]
I think this is a 4.9 release priority, so I'm adding the label to make it show up that way in the weekly newsletter.

/priority important-soon

File: enhancements/external-control-plane-topology.md [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+- Describe how the `controlPlaneTopology` field will be set for an external control plane deployment.
+
+## Proposal
+
+- Add `External` as an additional option to the `TopologyMode` type for `Infrastructure` 
+
+### Implementation Details
+
+Currently `External` only makes sense as a mode for the ControlPlaneTopology, not the InfrastructureTopology, but
```
Comment (eranco74) [inline enhancements/external-control-plane-topology.md:75]: Who will set this option and when?
Currently `HighlyAvailable` is the default `ControlPlaneTopology` and `SingleReplica` is set by the OpenShift installer according to the controlPlane replicas in the install-config.yaml.

File: enhancements/external-control-plane-topology.md [hunk 2]
```diff
@@ -0,0 +1,82 @@
+---
+title: External Control Plane Topology
+authors:
+  - "@csrwng"
+reviewers:
+  - "@derekwaynecarr"
+  - "@ironcladlou"
+  - "@enxebre"
+  - "@sjenning"
```
Comment (sttts) [inline enhancements/external-control-plane-topology.md:9]: why does this lack the leads of control-plane components?

File: enhancements/external-control-plane-topology.md [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+`IBMCloud` to distinguish it from other OCP installations with traditional
+control plane topology.
+
+More recently, an orthogonal field was [added to the Infrastructure resource](https://github.com/openshift/enhancements/blob/master/enhancements/single-node/cluster-high-availability-mode-api.md) to indicate
+the type of control plane topology. Current supported values for the `controlPlaneTopology` field
+are `HighlyAvailable` and `SingleReplica`.
+
+This enhancement proposes adding a third possible value to the control plane topology
+field: `External`. A value of `External` in this field indicates that control plane components
+such as Kube API server, Kube Controller Manager, and Kube Scheduler are running outside
```
Comment (deads2k) [inline enhancements/external-control-plane-topology.md:45]: and etcd?

File: enhancements/external-control-plane-topology.md [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+### Non-Goals
+
+- Provide a design for running OCP with an external control plane.
+
+- Describe how the `controlPlaneTopology` field will be set for an external control plane deployment.
+
+## Proposal
+
+- Add `External` as an additional option to the `TopologyMode` type for `Infrastructure`
```
[the remark this answers is not in the document]
Comment (csrwng) [reply enhancements/external-control-plane-topology.md:71]: @eranco74 this is not going to take the place of cluster profiles. The cluster profile is for the CVO to decide which manifests to include/exclude or to set appropriate node selectors for operators. The controlPlaneTopology value is for components to adjust their behavior once they're running.

File: enhancements/external-control-plane-topology.md (hunk 1 above)
Comment (csrwng) [reply enhancements/external-control-plane-topology.md:75]: A deployment with external control plane topology does not use the openshift installer, it uses an alternate method of installation (see https://github.com/openshift/hypershift or https://github.com/openshift/ibm-roks-toolkit). This alternate installer is responsible for setting the appropriate value for this field.

File: enhancements/external-control-plane-topology.md (hunk 2 above)
Comment (csrwng) [reply enhancements/external-control-plane-topology.md:9]: Sorry @sttts no reason at all. I added folks from hypershift by default but yes it really should include folks from various teams across OpenShift. Will update.

File: enhancements/external-control-plane-topology.md (hunk 3 above)
Comment (csrwng) [reply enhancements/external-control-plane-topology.md:45]: Yes, will update

**Comment by csrwng** [comment]
/cc @mfojtik @s-urbaniak @spadgett @dmage @Miciah 
Please let me know if there's additional folks that should review this.

File: enhancements/external-control-plane-topology.md (hunk 1 above)
Comment (csrwng) [reply enhancements/external-control-plane-topology.md:83]: Added text to explain

File: enhancements/external-control-plane-topology.md [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+is running on. IBM Cloud will soon support IPI/UPI installation. Thus, having a platform of `IBMCloud`
+will not imply an external control plane. Hypershift will bring support for external control planes
+to existing platforms such as AWS.
+
+### Goals
+
+- Allow expressing a new type of control plane topology in the Infrastructure resource inside an
+  OCP cluster.
+
+- Provide operators/components that change their behavior based on whether the control plane is external or
```
Comment (mrunalp) [inline enhancements/external-control-plane-topology.md:68]: Are the list of components that need to be changed captured somewhere?

File: enhancements/external-control-plane-topology.md [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+  self hosted, a clear indicator of what mode they're running in.
+
+### Non-Goals
+
+- Provide a design for running OCP with an external control plane.
+
+- Describe how the `controlPlaneTopology` field will be set for an external control plane deployment.
```
Comment (wking) [inline enhancements/external-control-plane-topology.md:75]: This enhancement is talking about the value it will be set to.  Are you saying that we will not specify the actor which sets the value?

**Comment by dhellmann** [comment]
There's a bug in the lint script that means the fact that this enhancement is missing a bunch of sections isn't caught by the automation. That bug is being fixed in #752.

The output of the script when I run it locally is:

```
enhancements/external-control-plane-topology.md missing "### User Stories"
enhancements/external-control-plane-topology.md missing "### Risks and Mitigations"
enhancements/external-control-plane-topology.md missing "## Design Details"
enhancements/external-control-plane-topology.md missing "### Test Plan"
enhancements/external-control-plane-topology.md missing "### Graduation Criteria"
enhancements/external-control-plane-topology.md missing "#### Dev Preview -> Tech Preview"
enhancements/external-control-plane-topology.md missing "#### Tech Preview -> GA"
enhancements/external-control-plane-topology.md missing "#### Removing a deprecated feature"
enhancements/external-control-plane-topology.md missing "### Upgrade / Downgrade Strategy"
enhancements/external-control-plane-topology.md missing "### Version Skew Strategy"
enhancements/external-control-plane-topology.md missing "## Implementation History"
enhancements/external-control-plane-topology.md missing "## Drawbacks"
enhancements/external-control-plane-topology.md missing "## Alternatives"
```

File: enhancements/external-control-plane-topology.md [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+both use the same type (`TopologyMode`).
+
+The Hypershift or IBM ROKS installer is responsible for setting this value when bootstrapping a new hosted control
+plane OpenShift cluster.
+
+#### Component Impact
+
+Existing components such as Console and Monitoring that use platform type of `IBMCloud` to modify their behavior
+for external control planes would need to switch to `controlPlaneTopology` to determine whether the control plane
+is external.
```
Comment (wking) [inline enhancements/external-control-plane-topology.md:93]: Is the test plan for this "CI jobs with an external control plane pass e2e without blowing up"?  Do we have CI coverage for things like whatever console twiddles based on this (looks like maybe just [dashboard availability][1])?  Or is the expectation that we make a best-effort attempt to transition existing `IBMCloud` consumers and then fix any we miss (or which slip in later) as we stumble across them?

[1]: https://github.com/openshift/console/blob/b7ff8039b6971e5ed30a5d3f4ecfbefce6d617a9/frontend/packages/console-app/src/plugin.tsx#L125-L140

**Comment by wking** [comment]
Now that #752 has landed, checking vs. this PR:

/test markdownlint

**Comment by wking** [comment]
openshift/release#18081 should remove the memory limit we were overshooting.

/retest

File: enhancements/external-control-plane-topology.md (hunk 5 above)
Comment (csrwng) [reply enhancements/external-control-plane-topology.md:68]: Currently it's components that have been modified to work with the IBMCloud platform type, will add a list with corresponding PRs

**Comment by csrwng** [comment]
@dhellmann the lint job passed, should I have seen an error about the missing sections?

File: enhancements/external-control-plane-topology.md (hunk 7 above)
Comment (csrwng) [reply enhancements/external-control-plane-topology.md:93]: Yes, I didn't think it was relevant to explain that Hypershift or IBM ROKS toolkit do their own bootstrapping and will set the value appropriately. But if you think it's relevant I can add it to the doc.

File: enhancements/external-control-plane-topology.md (hunk 6 above)
Comment (csrwng) [reply enhancements/external-control-plane-topology.md:75]: Yes, I didn't think it was relevant to explain that Hypershift or IBM ROKS toolkit do their own bootstrapping and will set the value appropriately. But if you think it's relevant I can add it to the doc.

File: enhancements/external-control-plane-topology.md (hunk 7 above)
Comment (csrwng) [reply enhancements/external-control-plane-topology.md:93]: The test plan for IBM Cloud is that QE on their side will verify that we have not regressed in the tweaks that have been made based on the `IBMCloud` platform. On our side, we don't yet have automated CI for Hypershift setups, but it's something that we're working towards.

**Comment by dhellmann** [comment]
> @dhellmann the lint job passed, should I have seen an error about the missing sections?

The lint job uses a script in the repo, so you'll need to rebase to see the errors.

**Comment by wking** [comment]
#755 landed.

/test markdownlint

**Comment by wking** [comment]
#756 has landed.

/test markdownlint

**Comment by wking** [comment]
Prow didn't notice.

/test markdownlint

**Comment by wking** [comment]
[Hooray][1] (I guess ;):

```
INFO[2021-04-27T16:40:00Z] enhancements/external-control-plane-topology.md missing "### User Stories" 
INFO[2021-04-27T16:40:00Z] enhancements/external-control-plane-topology.md missing "### Risks and Mitigations" 
INFO[2021-04-27T16:40:00Z] enhancements/external-control-plane-topology.md missing "## Design Details" 
...
```

[1]: https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_enhancements/744/pull-ci-openshift-enhancements-master-markdownlint/1387082860329963520#1:build-log.txt%3A20

**Comment by csrwng** [comment]
Will update later today :)

File: enhancements/external-control-plane-topology.md (hunk 6 above)
Comment (wking) [reply enhancements/external-control-plane-topology.md:75]: I'm fine leaving it out.  But maybe rephrase this line to:

> - Describe the mechanics of populating `controlPlaneTopology` the field.  This enhancement only introduces the new value.

or some such to make it more clear what portion is out of scope.

File: enhancements/external-control-plane-topology.md (hunk 5 above)
Comment (derekwaynecarr) [reply enhancements/external-control-plane-topology.md:68]: @mrunalp the list of components are useful, but really this is more saying 'there will never be a role=master' host in this cluster where set to `External`

File: enhancements/external-control-plane-topology.md (hunk 6 above)
Comment (derekwaynecarr) [reply enhancements/external-control-plane-topology.md:75]: @wking for clarification, the externalized control plane manager is responsible for projecting this value into the end-user cluster data plane.

**Review by derekwaynecarr — APPROVED** [review]
It looks like all comments were addressed, and this is needed to differentiate Standalone OCP on IBM Cloud from Managed Control Planes on IBM Cloud (or other infrastructures).

/approve
/lgtm

**Comment by dhellmann** [comment]
/hold

The lint job is failing because this document is missing a bunch of the required headers from the template. Please restore those sections and put something like `N/A` if they do not apply to this design.

```
+ hack/template-lint.sh
enhancements/external-control-plane-topology.md missing "### User Stories"
enhancements/external-control-plane-topology.md missing "### Risks and Mitigations"
enhancements/external-control-plane-topology.md missing "## Design Details"
enhancements/external-control-plane-topology.md missing "### Test Plan"
enhancements/external-control-plane-topology.md missing "### Graduation Criteria"
enhancements/external-control-plane-topology.md missing "#### Dev Preview -> Tech Preview"
enhancements/external-control-plane-topology.md missing "#### Tech Preview -> GA"
enhancements/external-control-plane-topology.md missing "#### Removing a deprecated feature"
enhancements/external-control-plane-topology.md missing "### Upgrade / Downgrade Strategy"
enhancements/external-control-plane-topology.md missing "### Version Skew Strategy"
enhancements/external-control-plane-topology.md missing "## Implementation History"
enhancements/external-control-plane-topology.md missing "## Drawbacks"
enhancements/external-control-plane-topology.md missing "## Alternatives"
```

File: enhancements/external-control-plane-topology.md (hunk 7 above)
Comment (s-urbaniak) [inline enhancements/external-control-plane-topology.md:93]: cc @simonpasquier

**Comment by csrwng** [comment]
@dhellmann I've added the missing sections
